### PR TITLE
Feature/profast fin model

### PR DESCRIPTION
**Pull request by jaredthomas68**
# Add a ProFAST financial model to HOPP custom financial

This PR adds a basic ProFAST model to the HOPP custom financial model so all technologies can be modeled with ProFAST inside HOPP instead of using the PySAM financial models. While an effort has been made to align parameter names with the names in PySAM, I was not able to find corresponding parameters for all the profast parameters in PySAM.Singleowner.

## Impacted areas of the software

```.github/workflows/ci.yml```
```.readthedocs.yaml```
```README.md```
```docs/requirements.txt```
```hopp/simulation/hopp_interface.py```
```hopp/simulation/hybrid_simulation.py```
```hopp/simulation/technologies/battery/battery.py```
```hopp/simulation/technologies/battery/battery_stateless.py```
```hopp/simulation/technologies/csp/csp_plant.py```
```hopp/simulation/technologies/dispatch/hybrid_dispatch_builder_solver.py```
```hopp/simulation/technologies/financial/custom_financial_model.py```
```hopp/simulation/technologies/grid.py```
```hopp/simulation/technologies/power_source.py```
```hopp/simulation/technologies/pv/detailed_pv_plant.py```
```hopp/simulation/technologies/pv/pv_plant.py```
```hopp/simulation/technologies/sites/site_info.py```
```hopp/simulation/technologies/wave/mhk_wave_plant.py```
```hopp/simulation/technologies/wind/wind_plant.py```
```pyproject.toml```
```tests/hopp/test_battery_dispatch.py```
```tests/hopp/test_custom_financial.py```
```tests/hopp/test_dispatch.py```
```tests/hopp/test_hybrid.py```
```tests/hopp/test_wave.py```
```tests/hopp/utils.py```

## Test results, if applicable

I changed the test value of two tests. One of them was off by less than 0.1% and I could not figure out what changed. The other required a change due to an added technology to the test.

### CODE REVIEW

**Comment by jaredthomas68** [comment]
I'm working on fixing the failed tests

**Comment by jaredthomas68** [comment]
The custom financial model in hopp already had methods for NPV and cash flow calculations. I would like to override those with the calculations from profast, but right now I only set up profast to provide LCOE (both nominal and real). I would like to hear others' thoughts on if we should move all calculations in custom financial to profast. This also raises the question of if we should just make a third financial model for profast so we leave custom alone. That would be a larger lift. Switching custom financial to fully profast would be fairly simple except that many tests would fail.

File: hopp/simulation/technologies/financial/custom_financial_model.py [hunk 1]
```diff
@@ -228,9 +252,155 @@ def execute(self, n=0):
...
+            "installation cost",
+            {
+                "value": 0,
+                "depr type": "Straight line",
+                "depr period": 4,
+                "depreciable": False,
+            },
+        )
+        pf.set_params("demand rampup", 0)
+        pf.set_params("long term utilization", 1)  # TODO should use utilization
```
Comment (kbrunik) [inline hopp/simulation/technologies/financial/custom_financial_model.py:341]: Does it make sense to change this to the capacity factor and use the rated capacity for the `set_param("capacity")`?

File: hopp/simulation/technologies/financial/custom_financial_model.py [hunk 2]
```diff
@@ -228,9 +252,155 @@ def execute(self, n=0):
...
+            nominal_discount_rate,
+        )
+        
+        pf.set_params(
+            "debt equity ratio of initial financing",
+            (
+                self.value('debt_percent')
+                / (100 - self.value('debt_percent'))
+            ),
+        )  # TODO this may not be put in right
```
Comment (kbrunik) [inline hopp/simulation/technologies/financial/custom_financial_model.py:379]: Maybe we can get @benjaminkee to weigh in here?

File: pyproject.toml [hunk 3]
```diff
@@ -47,6 +47,7 @@ dependencies = [
     "openpyxl",
     "attrs",
     "utm",
+    "pyyaml-include"
```
Comment (kbrunik) [inline pyproject.toml:50]: We have the load_yaml functionality in hopp, I don't think this is necessary? It doesn't look like it's used anywhere in this PR

File: hopp/simulation/technologies/battery/battery_stateless.py [hunk 4]
```diff
@@ -64,6 +64,7 @@ class BatteryStatelessConfig(BaseClass):
     maximum_SOC: float = field(default=90, validator=range_val(0, 100))
     initial_SOC: float = field(default=10, validator=range_val(0, 100))
     fin_model: Union[dict, CustomFinancialModel] = field(default=None)
```
Comment (kbrunik) [inline hopp/simulation/technologies/battery/battery_stateless.py:66]: Let's update this to 
```
fin_model: Optional[Union[str, dict, FinancialModelType]] = field(default=None)
```

File: tests/hopp/test_hybrid.py [hunk 5]
```diff
@@ -2,6 +2,7 @@
 from copy import deepcopy
 
 from pytest import approx, fixture, raises
+import pytest
```
Comment (kbrunik) [inline tests/hopp/test_hybrid.py:5]: Not sure we need this import

File: hopp/simulation/technologies/financial/custom_financial_model.py (hunk 2 above)
Comment (benjaminkee) [reply hopp/simulation/technologies/financial/custom_financial_model.py:379]: This looks correct. ProFAST uses debt to equity ratio, so if debt is 60 and equity is 40, then debt to equity is 1.5.

File: README.md [hunk 6]
```diff
@@ -44,6 +44,10 @@ solar and storage.
 
 4. Install HOPP and its dependencies:
 
+    ```bash
+    pip install ProFAST@git+https://github.com/NREL/ProFAST.git
```
[the remark this answers is not in the document]
Comment (jaredthomas68) [reply README.md]: Agreed. Updated

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -6,31 +6,47 @@ jobs:
...
           python-version: ${{ matrix.python-version }}
+          channels: conda-forge
+          activate-environment: hopp-test-${{ matrix.python-version }}
       - name: Install dependencies
         env:
           SKLEARN_ALLOW_DEPRECATED_SKLEARN_PACKAGE_INSTALL: True
         run: |
           sudo apt-get update && sudo apt-get install -y libglpk-dev glpk-utils coinor-cbc
           python -m pip install --upgrade pip
+          pip install ProFAST@git+https://github.com/NREL/ProFAST.git
```
[the remark this answers is not in the document]
Comment (jaredthomas68) [reply .github/workflows/ci.yml]: Done

File: docs/requirements.txt [hunk 8]
```diff
@@ -0,0 +1 @@
+ProFAST@git+https://github.com/NREL/ProFAST.git
```
[the remark this answers is not in the document]
Comment (jaredthomas68) [reply docs/requirements.txt]: Done

File: hopp/simulation/hybrid_simulation.py [hunk 9]
```diff
@@ -374,6 +374,7 @@ def set_om_costs(self, pv_om_per_kw=None, wind_om_per_kw=None,
         if self.battery:
             if battery_om_per_kw:
                 self.battery.om_capacity = battery_om_per_kw
```
[the remark this answers is not in the document]
Comment (jaredthomas68) [reply hopp/simulation/hybrid_simulation.py:376]: Good question. It looked like in some places the om_capacity may be used even for battery_om_per_kw. These are redundant names from my perspective that already existed in the code. I'm open to suggestions.

File: hopp/simulation/technologies/financial/custom_financial_model.py (hunk 2 above)
Comment (jaredthomas68) [reply hopp/simulation/technologies/financial/custom_financial_model.py]: Comment removed

File: pyproject.toml (hunk 3 above)
Comment (jaredthomas68) [reply pyproject.toml]: we do indeed have an include loader set up for our `load_yaml` function. I have removed pyyaml-include.

File: hopp/simulation/technologies/battery/battery_stateless.py (hunk 4 above)
Comment (jaredthomas68) [reply hopp/simulation/technologies/battery/battery_stateless.py]: Done. I also did the same for `battery.py`

File: tests/hopp/test_hybrid.py (hunk 5 above)
Comment (jaredthomas68) [reply tests/hopp/test_hybrid.py:5]: The `import pytest` is required for using the `subtests` fixture.

File: hopp/simulation/technologies/financial/custom_financial_model.py (hunk 1 above)
Comment (jaredthomas68) [reply hopp/simulation/technologies/financial/custom_financial_model.py:341]: I'm inclined to keep it as is. Using capacity factor and nameplate capacity is essentially having ProFAST make the same calculations we already did to get the actual capacity. I'm open to discussion here, but let's make it a separate issue/PR.

File: hopp/simulation/technologies/csp/csp_plant.py [hunk 10]
```diff
@@ -125,7 +125,7 @@ def __attrs_post_init__(self):
             raise AttributeError("Financial model must be set in `config.fin_model`")
 
         if isinstance(self.config.fin_model, dict):
-            financial_model = CustomFinancialModel(self.config.fin_model)
+            financial_model = CustomFinancialModel(self.config.fin_model, name=self.config.name)
```
[the remark this answers is not in the document]
Comment (jaredthomas68) [reply hopp/simulation/technologies/csp/csp_plant.py:128]: That is already there. Can you explain more if what you are suggesting is not there already?

File: hopp/simulation/technologies/financial/custom_financial_model.py (hunk 1 above)
Comment (kbrunik) [reply hopp/simulation/technologies/financial/custom_financial_model.py:341]: I think it makes sense to just make an issue about it, if we include multi-year simulations we will probably want to be able to add the dictionaries that include performance over time.

File: hopp/simulation/hybrid_simulation.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (kbrunik) [reply hopp/simulation/hybrid_simulation.py:376]: @dguittet is there a particular reason on the PySAM side that the battery has a parameter `om_batt_capacity_cost` and that this should be used over `om_capacity`? And that is that important in the custom financial model or more of a legacy effect?

File: hopp/simulation/technologies/financial/custom_financial_model.py [hunk 11]
```diff
@@ -228,9 +252,155 @@ def execute(self, n=0):
...
+                "capacity",
+                max([1E-6, self.value("batt_annual_discharge_energy")[0]/365.0]),
+            )  # kWh/day
+        else:
+            pf.set_params(
+                "capacity",
+                max([1E-6, self.value("annual_energy_kwh")/365.0]),
+            )  # kWh/day
+
+        pf.set_params("maintenance", {"value": self.o_and_m_cost(), "escalation": gen_inflation})
```
[the remark this answers is not in the document]
Comment (jaredthomas68) [reply hopp/simulation/technologies/financial/custom_financial_model.py:311]: I agree, but I consider the o_and_m_cost to be a separate issue as it impacts all financial models.

File: hopp/simulation/technologies/financial/custom_financial_model.py [hunk 12]
```diff
@@ -67,19 +67,40 @@ class SystemCosts(FinancialData):
...
-    ppa_escalation: float = field(default=1)
+    ppa_price_input: list = field(default=None) # cents/kWh
+    ppa_escalation: float = field(default=1) # percent (%)
     ppa_multiplier_model: float = field(default=None)
     dispatch_factors_ts: Sequence = field(default=(0,))
 
 
 @define
 class FinancialParameters(FinancialData):
+    """FinancialParameters
```
[the remark this answers is not in the document]
Comment (jaredthomas68) [reply hopp/simulation/technologies/financial/custom_financial_model.py]: Good point. Done.

File: hopp/simulation/technologies/financial/custom_financial_model.py [hunk 13]
```diff
@@ -67,19 +67,40 @@ class SystemCosts(FinancialData):
 
 @define
 class Revenue(FinancialData):
```
[the remark this answers is not in the document]
Comment (jaredthomas68) [reply hopp/simulation/technologies/financial/custom_financial_model.py:69]: Done

File: hopp/simulation/technologies/financial/custom_financial_model.py [hunk 14]
```diff
@@ -228,9 +252,155 @@ def execute(self, n=0):
...
+            )  # kWh/day
+        else:
+            pf.set_params(
+                "capacity",
+                max([1E-6, self.value("annual_energy_kwh")/365.0]),
+            )  # kWh/day
+
+        pf.set_params("maintenance", {"value": self.o_and_m_cost(), "escalation": gen_inflation})
+
+        # pf.add_fixed_cost(
```
[the remark this answers is not in the document]
Comment (johnjasa) [reply hopp/simulation/technologies/financial/custom_financial_model.py:346]: Briefly looking at ProFAST examples, I see they include both `set_params` and `add_fixed_cost` calls in the same example. I think keeping the `set_params` calls and removing the commented code here is a good path forward assuming it mirrors the expected PF usage.

**Review by johnjasa — COMMENTED** [review]
Looking good on my end, just a quick weigh-in comment! I'll avoid approving now; instead will wait for @kbrunik or others who have done a more complete review.

**Review by kbrunik — APPROVED** [review]
I'm happy with where this PR is at, still more work to do overall but I think this one is looking good to go!

**Comment by johnjasa** [comment]
Thanks @kbrunik! Is this ready to be merged in @jaredthomas68 or are you planning any further modifications? Could you please make issues for any upcoming changes that need to happen that are not part of this PR?

File: hopp/simulation/hybrid_simulation.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (jaredthomas68) [reply hopp/simulation/hybrid_simulation.py:376]: I have submitted an issue for this discussion to continue: #401

File: hopp/simulation/technologies/financial/custom_financial_model.py (hunk 1 above)
Comment (jaredthomas68) [reply hopp/simulation/technologies/financial/custom_financial_model.py:366]: I have created an issue to continue this discussion: #402

File: hopp/simulation/technologies/financial/custom_financial_model.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (jaredthomas68) [reply hopp/simulation/technologies/financial/custom_financial_model.py:344]: I'm going to go ahead and merge, but we can continue this discussion in PR #387

**Comment by jaredthomas68** [comment]
> Thanks @kbrunik! Is this ready to be merged in @jaredthomas68 or are you planning any further modifications? Could you please make issues for any upcoming changes that need to happen that are not part of this PR?

I have created or referenced existing issues for all outstanding conversations. I'm going ahead with merging this PR now.

**Pull request merged**